### PR TITLE
ddtrace/tracer: Golang tracer support for remote sampling rules 

### PR DESCRIPTION
**Pull request by yuanyuanzhao3**
DNMC-21

### What does this PR do?
Supports remote trace sampling rules. Also reports sampling rules (local or remote) through instrumentation telemetry for migration.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [X] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [X] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

**Comment by yuanyuanzhao3** [comment]
FYI, this is blocked behind a Golang tracer issue on output json of the sampling rule: https://dd.slack.com/archives/CKXKHAKFG/p1709736745814879

A new test thus fails without a fix to that problem.

**Comment by dianashevchenko** [comment]
@yuanyuanzhao3 hi there, have you looked into why the changes cause the increase in StartSpan time?

**Comment by yuanyuanzhao3** [comment]
> @yuanyuanzhao3 hi there, have you looked into why the changes cause the increase in StartSpan time?

Hi, I do not get a message that there was an increase in StartSpan time. The relevant email I got on performance is this:

> Benchmarks
> Benchmark execution time: 2024-03-06 18:02:13
> 
> Comparing candidate commit [0ec8d94](https://github.com/DataDog/dd-trace-go/commit/0ec8d94d4d35c2bd1b460208d0da4a14b438f076) in PR branch yuanyuan.zhao/remote-resource-tag-sampling-rules with baseline commit [63d7047](https://github.com/DataDog/dd-trace-go/commit/63d7047cfd4d8e3d078dc5460fb385e544c2eb1d) in branch main.
> 
> Found 0 performance improvements and 0 performance regressions! Performance is the same for 41 metrics, 0 unstable metrics.

Can you let me know where that StartSpan increase was reported? Thanks.

**Comment by dianashevchenko** [comment]
I am referring to the BenchmarkStartSpan, it was previously reported by the benchmarking bot. I see that the message is now edited and is no longer a case(you can see it in the history of the message ) With that said, I would recommend double checking just in case.
image

**Comment by yuanyuanzhao3** [comment]
> I am referring to the BenchmarkStartSpan, it was previously reported by the benchmarking bot. I see that the message is now edited and is no longer a case(you can see it in the history of the message ) With that said, I would recommend double checking just in case. image

Sounds good. I'll definitely check.

Also, it seems that there are PRs on the json marshaling change merged? Is it ready for me to pull in prod and update this PR? Just want to know so I go over it and check the benchmark together.

File: ddtrace/tracer/rules_sampler.go [hunk 1]
```diff
@@ -307,6 +343,18 @@ func (rs *traceRulesSampler) enabled() bool {
...
+func Equals(a, b []SamplingRule) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, r := range a {
+		if !r.Equals(&b[i]) {
+			return false
+		}
+	}
+	return true
```
Comment (dianashevchenko) [inline ddtrace/tracer/rules_sampler.go:355]: Just a question: is permutation a risk? Admittedly, it wouldn't change the end result however

File: ddtrace/tracer/telemetry.go [hunk 2]
```diff
@@ -94,6 +95,7 @@ func startTelemetry(c *config) {
 		telemetryConfigs = append(telemetryConfigs,
 			telemetry.Configuration{Name: fmt.Sprintf("sr_%s_(%s)_(%s)", rule.ruleType.String(), service, name),
 				Value: fmt.Sprintf("rate:%f_maxPerSecond:%f", rule.Rate, rule.MaxPerSecond)})
+		// TODO: encode provenance
```
Comment (dianashevchenko) [inline ddtrace/tracer/telemetry.go:98]: Does it still need to be done?

**Review by dianashevchenko — CHANGES_REQUESTED** [review]
Overall, looks good to me. Is there testing supplementing this code on the system-tests side?

**Comment by yuanyuanzhao3** [comment]
> Overall, looks good to me. Is there testing supplementing this code on the system-tests side?

Yeah. Misha is getting someone from the Python team to do it.

File: ddtrace/tracer/remote_config_test.go [hunk 3]
```diff
@@ -92,6 +114,50 @@ func TestOnRemoteConfigUpdate(t *testing.T) {
...
 	})
 
+	t.Run("DD_TRACE_SAMPLING_RULES rate=0.1 and RC trace sampling rules rate = 0.2", func(t *testing.T) {
+		telemetryClient := new(telemetrytest.MockClient)
+		defer telemetry.MockGlobalClient(telemetryClient)()
+
+		t.Setenv("DD_TRACE_SAMPLING_RULES", `[{
+				"service": "my-service",
+				"name": "web.request",
+				"resource": "*",
```
[the remark this answers is not in the document]
Comment (yuanyuanzhao3) [reply ddtrace/tracer/remote_config_test.go:124]: thanks for the comment.

On a related note, we are going to deprecate the trace global rates and turn them into match-alls. Your comments made a point that we should keep an eye on the performance impact.

File: ddtrace/tracer/remote_config_test.go [hunk 4]
```diff
@@ -92,6 +114,50 @@ func TestOnRemoteConfigUpdate(t *testing.T) {
...
+		require.Equal(t, 0.2, s.Metrics[keyRulesSamplerAppliedRate])
+		// Spans not matching the rule gets the global rate, but not the local rule, which is no longer in effect
+		s = tracer.StartSpan("web.request").(*span)
+		s.Resource = "not_abc"
+		s.Finish()
+		require.Equal(t, 0.5, s.Metrics[keyRulesSamplerAppliedRate])
+
+		telemetryClient.AssertNumberOfCalls(t, "ConfigChange", 1)
+		//telemetryClient.AssertCalled(t, "ConfigChange", []telemetry.Configuration{{Name: "trace_sample_rules", Value: 0.1, Origin: ""}})
+	})
```
[the remark this answers is not in the document]
Comment (yuanyuanzhao3) [reply ddtrace/tracer/remote_config_test.go:159]: fixed.

File: ddtrace/tracer/rules_sampler.go (hunk 1 above)
Comment (yuanyuanzhao3) [reply ddtrace/tracer/rules_sampler.go:355]: It is not a risk. However, I think the method should have a name like `EqualsFalseNegative` to indicate that it is not a strict/accurate `Equals`.

Also added the comment to explain.

File: ddtrace/tracer/telemetry.go (hunk 2 above)
Comment (yuanyuanzhao3) [reply ddtrace/tracer/telemetry.go:98]: Removed.

**Review by dianashevchenko — APPROVED** [review]

**Pull request merged**